torch/functional.py

```diff
@@ -324,6 +325,122 @@ def parse_subscript(n: int) -> str:
 
     return _VF.einsum(equation, operands)  # type: ignore[attr-defined]
 
+# If pad specifier is not None and not a tensor, convert it to a tensor.
+#
+# Arguments:
+#
+#   pad_spec    The pad specifier argument to convert to a tensor
+#
+#   arg_name    Name of pad specifier argument, used for error messages
+#
+#   dtype       The expected data type of the pad specifier tensor
+#
+#   device      The device to use for the pad specifier tensor
+def _maybe_convert_pad_specifier(pad_spec, arg_name, dtype, device):
+    if pad_spec is None or torch.is_tensor(pad_spec):
+        return pad_spec
+
+    # A valid pad specifier argument is either a single number, a sequence,
+    # of a homogeneous sequence of sequences. All of these can be represented
+    # as a tensor.
+    pad_spec_tensor = torch.tensor(pad_spec, device=device)
+
+    # If the expected dtype is an integer, we need to throw an error if the
+    # pad specifier argument contains any float or complex types.
+    # TODO: It would be more performant to specify dtype in the torch.tensor
+    # call and then catch errors and warnings it throws.
+    if pad_spec_tensor.dtype != dtype:
+        def is_int(dtype):
+            return (not dtype.is_floating_point) and (not dtype.is_complex)
+
+        if is_int(dtype) and not is_int(pad_spec_tensor.dtype):
+            raise RuntimeError(f"Expected '{arg_name}' to be of integer type")
+
+        pad_spec_tensor = pad_spec_tensor.to(dtype)
+
+    return pad_spec_tensor
+
+# This method is defined in Python because pad specifier arguments, like
+# `pad_width` and `constant_values`, can be given as a scalar, a 1-D or 2-D
+# sequence, or a tensor. To support all these different types in
+# a native_functions.yaml declaration, we would need several different
+# declarations for each pad specifier argument. So for simplicity, we only
+# declare tensor pad specifier arguments in native_functions.yaml, and handle
+# the other acceptable types in Python
+def pad(input, pad_width, mode='constant', *, constant_values=None, out=None):
+    r"""Pad a tensor by extending the size of one or more dimensions. The new
+    elements are filled with values according to the chosen padding mode.
+    Padding can be added to the beginning of each dimension, called "before
+    padding", and/or the end of each dimension, called "after padding".
+
```

Comment:
Should this seealso torch.nn.functional pad with a brief explanation of how they're different? Similarly, should torch.nn.functional.pad seealso this with a note that this pad is consistent with NumPy?

Reply:
Good idea. I've added a note about this to both